core/src/test/java/org/bitcoinj/core/TransactionTest.java

```diff
@@ -669,100 +669,62 @@ static <T, R extends T> IntFunction<R[]> genericArray(IntFunction<T[]> arrayCrea
         return size -> (R[]) arrayCreator.apply(size);
     }
 
-    @Test
+    @Test(expected = BufferUnderflowException.class)
     public void parseTransactionWithHugeDeclaredInputsSize() {
-        Transaction tx = new HugeDeclaredSizeTransaction(true, false, false);
-        byte[] serializedTx = tx.serialize();
-        try {
-            Transaction.read(ByteBuffer.wrap(serializedTx));
-            fail("We expect BufferUnderflowException with the fixed code and OutOfMemoryError with the buggy code, so this is weird");
-        } catch (BufferUnderflowException e) {
-            //Expected, do nothing
-        }
-    }
-
-    @Test
+        VarInt zero = VarInt.of(0);
+        VarInt huge = VarInt.of(Integer.MAX_VALUE);
+        // construct non-segwit transaction with huge txin count
+        ByteBuffer buf = ByteBuffer.allocate(4 + huge.getSizeInBytes() + zero.getSizeInBytes() + 4);
+        writeInt32LE(1, buf); // version
+        huge.write(buf); // txin count, txins intentionally missing
+        zero.write(buf); // txout count, no txouts
+        writeInt32LE(0, buf); // lock_time
+        checkState(!buf.hasRemaining());
+        ((Buffer) buf).rewind();
+
+        // make sure this safely throws BufferUnderflowException rather than running out of memory
+        Transaction.read(buf);
```

Comment:
This is definitely the right approach, but it is hard for me to quickly review this at the moment. It might be helpful if you created a method (or maybe even a class) for creating invalid `ByteBuffer`s for testing. It might even have the same parameters as before (e.g. `hackInputsSize`)